### PR TITLE
Preferred and default aggregator support

### PR DESCRIPTION
**Pull request by edfelten**
* Add a chain parameter that can specify a default aggregator, settable as a chain init option or via a builtin contract call
* Add a preferred aggregator attribute to L2 accounts. This can be an aggregator address, or None to specify using the default aggregator. The attribute is initially None for all accounts, and can be modified via a builtin contract call.
* When charging fees to a tx, only reimburse the validator if the tx was submitted by the sender's preferred validator (or the default validator if their preferred validator is None.) Otherwise, don't charge the sender for the reimbursable portion of fees.
* Add an ArbAggregator builtin contract to allow getting and setting of preferred aggregators (settable only by the account itself) and the default aggregator (settable only by the chain owner or the current default aggregator).
* Add tests

This functionality prevents an attack where a malicious aggregator front-runs an honest aggregator's transaction submissions in order to capture the honest aggregator's reimbursements.

If a chain has a single aggregator that is used by most users by default, that should be set as the default aggregator, so that users don't have to configure anything before using that aggregator.

### CODE REVIEW

File: node_modules/.yarn-integrity [hunk 1]
```diff
@@ -0,0 +1,12 @@
+{
```
Comment (hkalodner) [inline node_modules/.yarn-integrity:1]: This file should be removed

File: yarn.lock [hunk 2]
```diff
@@ -0,0 +1,4 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (hkalodner) [inline yarn.lock:1]: This file should be removed

File: src/compile/miniconstants.rs [hunk 3]
```diff
@@ -19,6 +19,8 @@ pub fn init_constant_table() -> HashMap<String, Uint256> {
         ("Address_ArbosTest", 105),
         ("Address_ArbOwner", 107),
         ("Address_ArbGasInfo", 108),
+        ("Address_ArbAggregator", 109),
+        // address 110 reserved for EthBridge functionality
```
Comment (hkalodner) [inline src/compile/miniconstants.rs:23]: If we're reserving this, I think the idea was to make any real call to the L2 contract at that address revert

File: arb_os/gasAccounting.mini [hunk 4]
```diff
@@ -798,7 +808,12 @@ impure func payForTxAndL1Gas(
         fullAmount
     };
 
-    let payableToFee = fixedPoint_mulByUintTrunc(feeConfig.shareForNetFee, amount);
+    let payableToFee = xif(preferredAggregator == recipient) {
+        fixedPoint_mulByUintTrunc(feeConfig.shareForNetFee, amount)
+    } else {
+        amount
+    };
```
Comment (hkalodner) [inline arb_os/gasAccounting.mini:815]: I'm not sure I understand why we're only incorporating the network fee here if using the preferred aggregator. I assume is has to do with the adjustment we made if you're not using the preferred aggregator above, but this is confusing and needs some refactoring or documentation

**Review by hkalodner — CHANGES_REQUESTED** [review]

File: arb_os/chainParameters.mini [hunk 5]
```diff
@@ -29,6 +29,7 @@ type ChainParams = struct {
     stakingToken: address,           // ERC-20 token used for staking; or zero if using ETH
     owner: address,                  // owner of the chain, who has admin privileges (or zero if there is no owner)
     secondsPerSend: option<uint>,    // issue at most one send per this many seconds
+    defaultAggregator: address,      // default aggregator, for users who haven't specified one
```
[the remark this answers is not in the document]
Comment (edfelten) [reply arb_os/chainParameters.mini:32]: This uses zero as the default aggregator address.  (`option` doesn't work well with Solidity APIs.)

File: node_modules/.yarn-integrity (hunk 1 above)
Comment (edfelten) [reply node_modules/.yarn-integrity:1]: Removed

File: src/compile/miniconstants.rs (hunk 3 above)
Comment (edfelten) [reply src/compile/miniconstants.rs:23]: Added code to do this.

File: yarn.lock (hunk 2 above)
Comment (edfelten) [reply yarn.lock:1]: Done.

File: arb_os/gasAccounting.mini (hunk 4 above)
Comment (edfelten) [reply arb_os/gasAccounting.mini:815]: Added comments to explain this.

**Review by hkalodner — APPROVED** [review]
LGTM

**Pull request merged**